toolkit/tools/internal/retry/retry.go

```diff
@@ -4,13 +4,29 @@
 package retry
 
 import (
+	"math"
 	"time"
 )
 
+func calcDelay(retryCount int, sleep time.Duration, backoffExp float64) time.Duration {
+	// Calculate an exponential backoff.
+	// The formula is: sleep * (backoffExp ^ retryCount)
+	// For example, if sleep = 1 second, backoffExp = 2.0, retryCount = 3
+	// then the delay will be 1 * (2 ^ 3) = 8 seconds.
+	expRetry := math.Pow(float64(retryCount), backoffExp)
+	return time.Duration(expRetry * float64(sleep))
+}
+
 // Run runs function up to attempts times, waiting i * sleep duration before each i-th attempt.
 func Run(function func() error, attempts int, sleep time.Duration) (err error) {
+	return RunWithExpBackoff(function, attempts, sleep, 1.0)
+}
+
+// Run runs function up to attempts times, waiting i * backoffMult * sleep duration before each i-th attempt.
+func RunWithExpBackoff(function func() error, attempts int, sleep time.Duration, backoffExp float64) (err error) {
 	for i := 0; i < attempts; i++ {
-		time.Sleep(time.Duration(i) * sleep)
+		// Calculate an exponential backoff.
+		time.Sleep(calcDelay(i, sleep, backoffExp))
```

Comment:
For `i = 0` won't this cause us to always sleep for `sleep` amount of time instead of immediately calling `function`?

Reply:
```suggestion
		if i > 0 {
			time.Sleep(calcDelay(i, sleep, backoffExp))
		}
```

Reply:
This is the old implementation, the calculation always results in sleep of zero for the 1st pass. I'll add a comment to that affect (and I validated this with a new test case).

Reply:
Reworked formulas, now has special casing for linear and failures=0